### DISCUSSION TITLE
Moor autogenerated primary key, value in null safety.

### DISCUSSION THREAD

**Discussion by techieasif** [simolus3/drift#1148] [Q&A] [answered]
Recently I have migrated my entire project to null safety.

This is One of my Transaction table in the project. 

```
class TBL_TXN_DCR_WW_CALLS extends Table {
  IntColumn get ID => integer().autoIncrement()();

  IntColumn get TXN_DCR_ID => integer().withDefault(Constant(0))();

  IntColumn get WW_PA_ID => integer().withDefault(Constant(0))();

  TextColumn get WW_PA_NAME =>
      text().withLength(min: 0, max: 250).withDefault(const Constant(""))();

  TextColumn get WW_CALL_TYPE =>
      text().withLength(min: 0, max: 250).withDefault(const Constant(""))();
}
```

The problem is before null safety migration was able to pass null value to primary key.
but now this is non nullable in data class of this table.

```
List<TBL_TXN_DCR_WW_CALLSData> wwDRCALLS = List.generate(
        workWiths.length,
        (index) => TBL_TXN_DCR_WW_CALLSData(
            ID: null, //error: The argument type 'Null' can't be assigned to the parameter type 'int'.
            WW_CALL_TYPE: _callType,
            TXN_DCR_ID: txnDcrId,
            WW_PA_ID: workWiths[index].paId,
            WW_PA_NAME: workWiths[index].paName));
    await this.dbDao.insertAll_TXN_WW_CALLS(wwDRCALLS);
```

i have tried to add nullable type in table definition as 

```
 IntColumn? get ID => integer().autoIncrement()();
```

 GENERATED DATA CLASS OF THIS TABLE:
```

class TBL_TXN_DCR_WW_CALLSData extends DataClass
    implements Insertable<TBL_TXN_DCR_WW_CALLSData> {
  final int ID; //NON Nullable

  ///this [WW_CALL_ID] will be the reference to save and get ww calls
  ///it will store DR id, chemist id, stokist to make relations with them.
  final int TXN_DCR_ID;
  final int WW_PA_ID;
  final String WW_PA_NAME;
  final String WW_CALL_TYPE;
  TBL_TXN_DCR_WW_CALLSData(
      {required this.ID, //REQUIRED
      required this.TXN_DCR_ID,
      required this.WW_PA_ID,
      required this.WW_PA_NAME,
      required this.WW_CALL_TYPE});
  factory TBL_TXN_DCR_WW_CALLSData.fromData(
      Map<String, dynamic> data, GeneratedDatabase db,
      {String? prefix}) {
    final effectivePrefix = prefix ?? '';
    final intType = db.typeSystem.forDartType<int>();
    final stringType = db.typeSystem.forDartType<String>();
    return TBL_TXN_DCR_WW_CALLSData(
      ID: intType.mapFromDatabaseResponse(data['${effectivePrefix}id'])!,
      TXN_DCR_ID: intType
          .mapFromDatabaseResponse(data['${effectivePrefix}txn_dcr_id'])!,
      WW_PA_ID:
          intType.mapFromDatabaseResponse(data['${effectivePrefix}ww_pa_id'])!,
      WW_PA_NAME: stringType
          .mapFromDatabaseResponse(data['${effectivePrefix}ww_pa_name'])!,
      WW_CALL_TYPE: stringType
          .mapFromDatabaseResponse(data['${effectivePrefix}ww_call_type'])!,
    );
  }

```
but it not worked.
and aslo primary key can't be nullable.

Please help me regarding this?

**Comment by davidmartos96** [comment, marked as answer]
moor generates an extra class along your table data class which is called (MyTable)Companion.
That is meant to be used for insertions or updates. The id would be optional when using the companion. So instead of creating a list of data classes try creating a list of companions

**Reply by techieasif** [reply]
Ok Thanks, is there any alternative possible? without converting it to companions.
Actually i have around 40 tables😅.

**Reply by simolus3** [reply]
Data classes are designed to hold a full row, with all of its values. Since the id is a (non-nullable) value, it started being required with null safety (it actually was `@required` before, it just wasn't a compile time error to not pass a value). If moor made it nullable, it would be very annoying to use `.id!` after each select, when `id` is actually known to be non-nullable in the database. So, the solution was to introduce another class for this purpose.

Migrating to companions is a bit easier with the generated `CompanionClass.insert` constructor, where all the fields that are truly required for inserts are marked as required. In your app, you can in most parts replace `YourDataClass` with `Insertable<YourDataClass>` to allow both data classes and companions.

Also, please don't tag everyone for a question.

**Reply by techieasif** [reply]
Thank you so much @simolus3 for your time.

I'll keep in mind about tagging next time.
